modules/stochastic_tools/src/samplers/AdaptiveImportanceSampler.C

```diff
@@ -0,0 +1,196 @@
+const std::vector<Real> &
+AdaptiveImportanceSampler::getInitialValues() const
+{
+  return _initial_values;
+}
+
+const int &
+AdaptiveImportanceSampler::getNumSamplesTrain() const
+{
+  return _num_samples_train;
+}
+
+const bool &
+AdaptiveImportanceSampler::getUseAbsoluteValue() const
+{
+  return _use_absolute_value;
+}
+
+const Real &
+AdaptiveImportanceSampler::getOutputLimit() const
+{
+  return _output_limit;
+}
+
+Real
+AdaptiveImportanceSampler::computeSample(dof_id_type /*row_index*/, dof_id_type col_index)
+{
+  TIME_SECTION(_perf_compute_sample);
+
+  if (_step <= _num_samples_train)
+  {
+    /* This is the importance distribution training step. Markov Chains are set up
+       to sample from the importance region or the failure region using the Metropolis
+       algorithm. Given that the previous sample resulted in a model failure, the next
+       sample is proposed such that it is very likely to result in a model failure as well.
+       The `initial_values` and `proposal_std` parameters provided by the user affects the
+       formation of the importance distribution. */
+    if (_step > 1 && col_index == 0 && _check_step != _step)
+    {
+      for (dof_id_type j = 0; j < _distributions.size(); ++j)
+        _prev_value[j] = Normal::quantile(
+            _distributions[j]->cdf(getReporterValue<std::vector<Real>>("inputs_reporter")[j]),
+            0,
+            1);
+      _acceptance_ratio = 0.0;
+      for (dof_id_type i = 0; i < _distributions.size(); ++i)
+        _acceptance_ratio += std::log(Normal::pdf(_prev_value[i], 0, 1)) -
+                             std::log(Normal::pdf(_inputs_sto[i][_inputs_sto[i].size() - 1], 0, 1));
+      if (_acceptance_ratio > std::log(getRand(_step)))
+      {
+        for (dof_id_type i = 0; i < _distributions.size(); ++i)
+          _inputs_sto[i].push_back(_prev_value[i]);
+      }
+      else
+      {
+        for (dof_id_type i = 0; i < _distributions.size(); ++i)
+          _inputs_sto[i].push_back(_inputs_sto[i][_inputs_sto[i].size() - 1]);
+      }
+    }
+    _check_step = _step;
+    _prev_value[col_index] =
+        Normal::quantile(getRand(_step),
+                         _inputs_sto[col_index][_inputs_sto[col_index].size() - 1],
+                         _proposal_std[col_index]);
+    return _distributions[col_index]->quantile(Normal::cdf(_prev_value[col_index], 0, 1));
+  }
+  else if (_step == _num_samples_train + 1)
+  {
+    /* This is the importance sampling step using the importance distribution created
+       in the previous step. Once the importance distribution is known, sampling from
+       it is similar to a regular Monte Carlo sampling. */
+    if (col_index == 0 && _check_step != _step)
+    {
+      for (dof_id_type i = 0; i < _distributions.size(); ++i)
+      {
+        _mean_sto[i] = StochasticTools::AdaptiveMonteCarloUtils::computeMEAN(_inputs_sto[i], 2);
+        _std_sto[i] = StochasticTools::AdaptiveMonteCarloUtils::computeSTD(_inputs_sto[i], 2);
```

Comment:
Seems like there will be a problem here if the number of columns is less than 3 because of that `start_index = 2`. Is it just returning 0 in this case?